content/posts/building-container-images-with-kubernetes.md

```diff
@@ -0,0 +1,190 @@
+
+## Building Container Images without Docker
+Or more specifically, building container images without the Docker **daemon** - no **DinD**, no Docker socket. But we still want to leverage K8s managed containers for CD, to include building and pushing container images. How do we do that?
+
+### Enter Kaniko
+Kaniko is a tool that is capable of building and pushing container images without the Docker daemon, but it does have one major drawback from a security perspective: to build anything useful/easily you must run as the `root` `USER` in the Kaniko container. 
+
+>[**From Kaniko's docs**](https://github.com/GoogleContainerTools/kaniko#security): If you have a minimal base image (SCRATCH or similar) that doesn't require permissions to unpack, and your Dockerfile doesn't execute any commands as the root user, you can run Kaniko without root permissions. It should be noted that Docker runs as root by default, so you still require (in a sense) privileges to use Kaniko.
+
+>You may be able to achieve the same default seccomp profile that Docker uses in your Pod by setting seccomp profiles with annotations on a PodSecurityPolicy to create or update security policies on your cluster. 
+
+As we already mentioned above, running as `root` is an attack vector that many consider to be an unacceptable security hole - but the use of Pod Security Policies will reduce the attack surface of the Kaniko container running as part of a K8s Pod and provides greater security than the Docker based approaches we have already dismissed.
+
+#### Basic Configuration for Kaniko with Pod Security Policies
+The following configuration assumes that you have a K8 cluster up and running, and that you have access to `kubectl` to add and modify K8s resources.
+
+1. **Container registry**: I recommend having a staging or sandbox container registry that Kaniko pushes images to, and a production registry that Kaniko does not have access to push to. A separately managed and secured CD job should be used to promote container images from staging to production once required tests/scans/policies have been successfully run against that container image. But of course the Kaniko container image itself (and any other container images used by CD jobs) should always be pulled from the production container registry.
+2. `PodSecurityPolicy`: Once the PodSecurityPolicy admission controller is enabled you will need at least 2 PodSecurityPolicies (acutally you will want to defined and apply all of your Pod Security Policies before enabling the admission controller, not doing so will prevent any pods from being created in the cluster):
```

Comment:
Typo on `(acutally`

Reply:
I guess Spell Right missed that one, thanks!